File: 백준/Silver/19941. 햄버거 분배/햄버거 분배.cc
#include <iostream>
#include <string>
#include <vector>
#include <cstdlib>
using namespace std;

int main(void){
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    int N, K;
    string str;
    vector<int> p, h;
    cin >> N >> K;
    cin >> str;
    for(int i = 0; i < N; i++){
        if(str[i] == 'P') p.push_back(i);
        else h.push_back(i);
    }

    int ans = 0, pIdx = 0, hIdx = 0;
    while(pIdx < p.size() && hIdx < h.size()){
        if(abs(p[pIdx] - h[hIdx]) <= K){
            pIdx++; hIdx++;
            ans++;
        }else{
            if(p[pIdx] < h[hIdx]) pIdx++;
            else hIdx++;
        }
    }
    cout << ans;
}

File: 백준/Silver/19941. 햄버거 분배/햄버거 분배.js
const input = require("fs")
    .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
    .toString()
    .trim()
    .split("\n");
    // .map((el) => el.split(" ").map(Number));

let vars = input[0].split(" ").map(Number);
let N = vars[0], K = vars[1];
let str = input[1];
let person = [];
let hamburger = [];

for(let i = 0; i < N; i++){
    if(str[i] == 'P') person.push(i);
    else hamburger.push(i);
}

let pIdx = 0, hIdx = 0, ans = 0;
while(pIdx < person.length && hIdx < hamburger.length){
    if(Math.abs(person[pIdx] - hamburger[hIdx]) <= K){
        pIdx++; hIdx++;
        ans++;
    }else{
        if(person[pIdx] < hamburger[hIdx]) pIdx++;
        else hIdx++;
    }
}
console.log(ans);